File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <iostream>
#include <vector>

using namespace std;

void getPermutation(vector<int> seq,vector<int> cur, vector<bool> isVisited, int m) {
	
	if (cur.size() == m) {
		for (int n : cur)
			cout << seq[n] << " ";
		cout << "\n";
	}
	else {
		

		for (int i = 0; i < seq.size(); i++) {

			if (isVisited[i] == true)
				continue;
			cur.push_back(i); isVisited[i] = true;
			getPermutation(seq,cur,isVisited,m);
			cur.pop_back(); isVisited[i] = false;
		}
	}
}



int main(void) {
	int n, m;
	cin >> n >> m;
	
	vector<int> seq;
	vector<bool> isVisited;
	for (int i = 0; i < n; i++) {
		seq.push_back(i + 1);
		isVisited.push_back(false);
	}
	

	vector<int> cur;
	getPermutation(seq, cur, isVisited, m);
	
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.Scanner;

public class Main {

	
	public static void getPermutation(int[] seq,int[] cur, boolean[] isVisited, int sizeOfCur, int m) {
		
		if (sizeOfCur == m) {
			for(int i = 0; i <sizeOfCur; i++)
				System.out.print(seq[cur[i]] + " ");
			System.out.println();
		}
		else {
			
			for (int i = 0; i < seq.length; i++) {
				
				if (isVisited[i] == true)
					continue;
				
				cur[sizeOfCur++] = i; 
				isVisited[i] = true;
				getPermutation(seq,cur,isVisited,sizeOfCur--,m);
				isVisited[i] = false;
				
			}
		}
	}



	public static void main(String args[]) {
		
		Scanner scanner = new Scanner(System.in);
		
		
		
		int n = scanner.nextInt();
		int m = scanner.nextInt();
		
		
		int[] seq = new int[n];
		boolean[] isVisited = new boolean[n];
		
	
		for (int i = 0; i < n; i++) {
			seq[i] = i+1;
			isVisited[i] = false;
		}
		

		int[]  cur = new int[m];
		getPermutation(seq, cur, isVisited,0, m);
		
	}
}